virtualmachine/gcp/util.go

```diff
@@ -435,3 +442,284 @@ func (svc *googleService) insertSSHKey() error {
+}
+
+// getRegionList gets the list of regions
+func (svc *googleService) getRegionList() ([]*googlecloud.Region, error) {
+	regionList, err := svc.service.Regions.List(svc.vm.Project).Do()
+	if err != nil {
+		return nil, err
+	}
+
+	return regionList.Items, nil
+}
+
+// reset resets a GCE instance.
+func (svc *googleService) reset() error {
+	instance, err := svc.getInstance()
+	if err != nil {
+		return err
+	}
+
+	if instance.Status != "RUNNING" {
+		return fmt.Errorf("instance %s is not in RUNNING status, "+
+			"cannot reset instance in %s status",
+			instance.Name, instance.Status)
+	}
+
+	op, err := svc.service.Instances.Reset(svc.vm.Project, svc.vm.Zone, svc.vm.Name).Do()
+	if err != nil {
+		return err
+	}
+
+	return svc.waitForOperationReady(op.Name)
+}
+
+// getDiskType gets details of a disk type
+func (svc *googleService) getDiskType(diskType string) (*googlecloud.DiskType, error) {
+	return svc.service.DiskTypes.Get(svc.vm.Project, svc.vm.Zone, diskType).Do()
+}
+
+// createDisk creates a new disk in GCE
+func (svc *googleService) createDisk(disk *Disk) error {
+	diskType, err := svc.getDiskType(disk.DiskType)
+	if err != nil {
+		return err
+	}
+
+	gDisk := &googlecloud.Disk{
+		Name:        disk.Name,
+		SizeGb:      int64(disk.DiskSizeGb),
+		Type:        diskType.SelfLink,
+		Description: disk.Description,
+	}
+
+	op, err := svc.service.Disks.Insert(svc.vm.Project, svc.vm.Zone, gDisk).Do()
+	if err != nil {
+		return err
+	}
+	return svc.waitForOperationReady(op.Name)
+}
+
+// attachDisk attaches a disk to an instance
+func (svc *googleService) attachDisk(disk *Disk) error {
+
+	gDisk, err := svc.getDisk(disk.Name)
+	if err != nil {
+		return err
+	}
+
+	gAttachedDisk := &googlecloud.AttachedDisk{
+		Source:     gDisk.SelfLink,
+		DeviceName: disk.Name,
+		AutoDelete: disk.AutoDelete,
+	}
+
+	op, err := svc.service.Instances.AttachDisk(svc.vm.Project, svc.vm.Zone,
+		svc.vm.Name, gAttachedDisk).Do()
+
+	if err != nil {
+		return err
+	}
+	return svc.waitForOperationReady(op.Name)
+}
+
+// detachDisk detaches a disk from an instance
+func (svc *googleService) detachDisk(disk *Disk) error {
+
+	var deviceName string
+
+	gDisk, err := svc.getDisk(disk.Name)
+	if err != nil {
+		return err
+	}
+	instance, err := svc.getInstance()
+	if err != nil {
+		return err
+	}
+
+	for _, attachedDisk := range instance.Disks {
+		if attachedDisk.Source == gDisk.SelfLink {
+			deviceName = attachedDisk.DeviceName
+			break
+		}
+	}
+
+	op, err := svc.service.Instances.DetachDisk(svc.vm.Project, svc.vm.Zone,
+		svc.vm.Name, deviceName).Do()
+
+	if err != nil {
+		return err
+	}
+	return svc.waitForOperationReady(op.Name)
+
+}
+
+// getFirewall gets details of a firewall
+func (svc *googleService) getFirewall() (*googlecloud.Firewall, error) {
+	return svc.service.Firewalls.Get(svc.vm.Project, svc.vm.Firewall).Do()
+}
+
+// addFirewallPorts adds new ports to a firewall
+func (svc *googleService) addFirewallPorts() error {
```

Comment:
We are adding given protocol:port to the existing rules of a given firewall (in other words, yes, we are extending the existing list). GCP takes care of de-duplication if any. Let me know in case of any issue.